### DISCUSSION TITLE
Problem with function login

### DISCUSSION THREAD

**Discussion by lokize** [PostgREST/postgrest#2172] [Q&A] [answered]
Hi guys, i have this problem when i try call endpoint /rpc/login:

{  "message": "Could not find the public.login() function in the schema cache",
  "hint": "If a new function was created in the database with this name and parameters, try reloading the schema cache."
}

but i created function login in public schema not in cache, please how to fix that?

**Comment by laurenceisla** [comment]
Hi! The error mentions that the function cannot be found in the [Schema Cache](https://postgrest.org/en/latest/schema_cache.html) that PostgREST uses to get metadata from the database. What may be happening is that you had PostgREST running when you created the function, causing the cache to be stale. Try [reloading it](https://postgrest.org/en/latest/schema_cache.html#schema-cache-reloading) to see if it works.

If it does not help, share the function definition to see how it was created.

**Comment by lokize** [comment]
thanks for helping

not work, my function login is:

```sql
CREATE OR REPLACE FUNCTION cache.login(email text, pass text)
 RETURNS basic_auth.jwt_token
 LANGUAGE plpgsql
 SECURITY DEFINER
AS $function$
declare
  _role name;
  result basic_auth.jwt_token;
begin
  -- check email and password
  select basic_auth.user_role(email, pass) into _role;
  if _role is null then
    raise invalid_password using message = 'invalid user or password';
  end if;

  select sign(
      row_to_json(r), 'reallyreallyreallyreallyreallyverysafe'
    ) as token
    from (
      select _role as role, login.email as email,
         extract(epoch from now())::integer + 60*60 as exp
    ) r
    into result;
  return result;
end;
$function$
;
```

function is created in my database: 

<img width="1412" alt="image" src="https://user-images.githubusercontent.com/19701630/154647797-33fd3d57-bf52-40be-b886-cbbb4e2ab0a6.png">

**Reply by wolfgangwalther** [reply]
The error message is `Could not find the public.login() function in the schema cache`.

"schema cache" is not meant as the `SCHEMA` with the name "cache" - but the cache of schemas that is kept by PostgREST internally.

> ```sql
> CREATE OR REPLACE FUNCTION cache.login(email text, pass text)
> ```

This is wrong. You need to create the function in the `public` schema.

**Reply by lokize** [reply]
<img width="831" alt="image" src="https://user-images.githubusercontent.com/19701630/154648106-276189d2-f885-4ab2-b0c7-3250acad5eb6.png">

**Reply by wolfgangwalther** [reply]
Please stop posting screenshots. It's enough to copy the error message you are getting.

---

I tell you "this is wrong" and the only thing you do is copy paste the **wrong** snippet into your editor and execute it? How is this supposed to work?

You need to change `cache.login` to `public.login` - and then run your **full** `CREATE OR REPLACE FUNCTION` statement again.

**Reply by lokize** [reply]
I'm sorry, I didn't mean to disturb the intention of posting the image, it was to better exemplify the problem, sorry!

**Comment by lokize** [comment, marked as answer]
i changed function restarted database and reloaded schema the error changed to:

{
  "message": "Could not find the public.login() function or the public.login function with a single unnamed json or jsonb parameter in the schema cache",
  "hint": "If a new function was created in the database with this name and parameters, try reloading the schema cache."
}

**Reply by wolfgangwalther** [reply]
How exactly does the request look like you're making? HTTP method, URI, Headers, Body.

**Reply by lokize** [reply]
POST in localhost:8081/rpc/login with email and pass in header, no body.

**Reply by wolfgangwalther** [reply]
You need to pass a `Content-Type: application/json` header and pass the arguments to the function in the body:

```json
{
  "email": "...",
  "pass": "..."
}
```

You can **not** pass function arguments in a header.

**Reply by lokize** [reply]
wow, this working, thanks man you really save my life, i have another problem when i try using token in my get request: 

{
  "message": "JWSError JWSInvalidSignature"
}

i put token in Auth Bearer.

**Reply by wolfgangwalther** [reply]
Did you set the same ~~token~~ secret you have in your login function (`reallyreallyreallyreallyreallyverysafe`) as the config option `jwt-secret`?

**Reply by lokize** [reply]
i used sql to put secret: alter role anon  in database cerescliente set pgrst.jwt_secret = (my code generated by jose in rsa.jwk.pub)

**Reply by wolfgangwalther** [reply]
`alter role anon in database` this is wrong. You should **not** set that on the `anon` user, but on the `authenticator` user (the one that postgrest connects with).

**Reply by lokize** [reply]
in conf file i need to add jwt-secret too? and my db-anon-role = (need to be authenticator?)

**Reply by wolfgangwalther** [reply]
You need to pass the secret to PostgREST. To do this you can do **one** of:
- set it in the config file
- set it on the authenticator role

`db-anon-role` should **not** be `authenticator`, but should remain at `anon`.

**Reply by lokize** [reply]
i changed only in database, i put secret do default documentation (really...), after generating token i try GET with header Authorization: Bearer mytoken and i have {
"message": "JWSError JWSInvalidSignature"
}

**Reply by lokize** [reply]
Please close this thread i fixed the problem doing the process from scratch a few times until I understood how it actually worked, thanks to the @wolfgangwalther for the help and patience. thank you!
